File: examples/03-interactions/go/update_with_start.go
package interactions

import (
	"context"

	enumspb "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/client"
)

// Update-with-Start atomically starts the Workflow (if not already running) and
// applies an Update in one round trip - the "add to cart, creating the cart if
// needed" pattern. The start half is a WithStartWorkflowOperation; the conflict
// policy USE_EXISTING reuses a running cart instead of erroring.
func AddItemOrCreateCart(ctx context.Context, c client.Client) (int, error) {
	startOp := c.NewWithStartWorkflowOperation(
		client.StartWorkflowOptions{
			ID:                       "cart-1001",
			TaskQueue:                "carts",
			WorkflowIDConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
		},
		CartWorkflow,
		"cart-1001",
	)

	handle, err := c.UpdateWithStartWorkflow(ctx, client.UpdateWithStartWorkflowOptions{
		StartWorkflowOperation: startOp,
		UpdateOptions: client.UpdateWorkflowOptions{
			UpdateName:   "addItem",
			Args:         []any{"lamp", 1},
			WaitForStage: client.WorkflowUpdateStageCompleted,
		},
	})
	if err != nil {
		return 0, err
	}

	var itemCount int
	if err := handle.Get(ctx, &itemCount); err != nil {
		return 0, err
	}
	return itemCount, nil
}

File: examples/03-interactions/python/update_with_start.py
from temporalio.client import (
    Client,
    WithStartWorkflowOperation,
    WorkflowUpdateStage,
)
from temporalio.common import WorkflowIDConflictPolicy


# Update-with-Start atomically starts the Workflow (if not already running) and
# applies an Update in one round trip - the "add to cart, creating the cart if
# needed" pattern. The start half is a WithStartWorkflowOperation; the conflict
# policy USE_EXISTING reuses a running cart instead of erroring.
async def add_item_or_create_cart(client: Client) -> int:
    start_op = WithStartWorkflowOperation(
        CartWorkflow.checkout,
        "cart-1001",
        id="cart-1001",
        task_queue="carts",
        id_conflict_policy=WorkflowIDConflictPolicy.USE_EXISTING,
    )

    handle = await client.start_update_with_start_workflow(
        CartWorkflow.add_item,
        args=["lamp", 1],
        start_workflow_operation=start_op,
        wait_for_stage=WorkflowUpdateStage.COMPLETED,
    )

    item_count: int = await handle.result()
    return item_count
